### PR TITLE
Adding test for HTTP/2 draining

### PR DESCRIPTION
**Pull request by zizhong**
@maskit @masaori335 can you please review?
@bryancall @zwoop I copied the test structure from hyper example code in MIT license. is it okay for us to use?

### CODE REVIEW

**Comment by maskit** [comment]
[approve ci]

From my understanding, copyright statement has to be kept even if it was MIT license, if you *copied* the example code.

**Comment by zizhong** [comment]
@maskit Do I also need to add the Apache license?

**Comment by maskit** [comment]
Yes, probably we should do that. Code that licensed under MIT should be able to sublicense under Apache license.

**Comment by maskit** [comment]
[approve ci]

**Comment by maskit** [comment]
It looks a bit confusing. Maybe someone read it as dual license, which users can choose one from them.

You should clarify that the original code **was** distributed under MIT license with the copyright notice.

@zwoop @bryancall Do you have good examples or references?

**Comment by maskit** [comment]
The test itself looks good to me. Only concern is availability and compatibility of `pkill` command. If possible, it should be done by calling some function provided by Python.

**Comment by maskit** [comment]
[approve ci]

**Comment by zizhong** [comment]
@maskit updated! can you review it again? Thanks!

**Comment by maskit** [comment]
[approve ci]

**Comment by maskit** [comment]
[approve ci autest]

**Comment by zizhong** [comment]
Not sure why the autest failed on this one. Works locally every time. I need to look into it more.

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci debian]

File: tests/gold_tests/h2/http2.test.py [hunk 1]
```diff
@@ -75,11 +76,13 @@
     'proxy.config.ssl.server.cipher_suite': 'ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384:AES128-GCM-SHA256:AES256-GCM-SHA384:ECDHE-RSA-RC4-SHA:ECDHE-RSA-AES128-SHA:ECDHE-RSA-AES256-SHA:RC4-SHA:RC4-MD5:AES128-SHA:AES256-SHA:DES-CBC3-SHA!SRP:!DSS:!PSK:!aNULL:!eNULL:!SSLv2',
     'proxy.config.http2.active_timeout_in': 3,
     'proxy.config.http2.max_concurrent_streams_in': 65535,
+    'proxy.config.stop.shutdown_timeout': 3,
 })
 ts.Setup.CopyAs('h2client.py', Test.RunDirectory)
```
Comment (dragon512) [inline tests/gold_tests/h2/http2.test.py]: All these CopyAs can be more terse with just a 
ts.Setup.Copy('h2client.py')

**Comment by zwoop** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
@zizhong Can you fix the merge conflict?

@zizhong @dragon512 Is this ready to land after the merge has been fixed?

**Comment by zizhong** [comment]
Let me check on the conflicts.
@bryancall Any comment on the dual licenses?

**Comment by zizhong** [comment]
I updated my branch and the conflict is gone. However, the test is not working for now.
I'll submit a small fix soon.

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by zwoop** [comment]
@zizhong What do we want to do with this PR ?

**Comment by zizhong** [comment]
@zwoop I'll update this and #3693 today.

**Comment by zizhong** [comment]
[approve ci autest]

**Comment by zizhong** [comment]
```Traceback (most recent call last):
  File "h2drain.py", line 26, in <module>
    import twisted.internet._sslverify as v
ModuleNotFoundError: No module named 'twisted'
```
, but I added `twisted` to `bootstrap.py`.
@zwoop @dragon512 what's the right way to tell the Autest env installing new python libs?

**Comment by bryancall** [comment]
approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
[approve ci autest]

**Comment by bryancall** [comment]
@zizhong The PR is failing in autest, I would like to get it merged or close it since it hasn't been updated in awhile.

**Comment by zizhong** [comment]
@bryancall these tests work perfectly locally, where I manually installed twisted. To get it passing the autests,  I would like to get some help to figure out what the correct way to install a python dep in autest. Do you happen to know the answer? Thanks!
cc @zwoop @dragon512 @maskit

**Comment by dragon512** [comment]
I have no issue with using twisted. It is network based python framework used by lots of people. I think the question for me at the moment is what is it being used for here and can this be done with the other tools we already have. @macisasandwich Can you look and see if this could be done with our existing tools you have been putting effort into?

**Comment by macisasandwich** [comment]
Right now we've got no H2 capabilities with microserver and it will probably be an undefined amount of time before I can get around to adding H2 and draining to the tools that we use. So we should add this one in.

**Comment by zwoop** [comment]
I feel reasonably strongly that we should not introduce Twisted into our system. Things are already complicated, and Twisted amplifies that by another magnitude.

**Comment by zwoop** [comment]
Also, not sure why you would need H2 capabilities in the micro server ? This is acting as an H2 client (I can only assume, I can not decipher the code), and we have no H2 to origin support already.

Can't we implement this as either just using curl commands, or perhaps some simpler, basic Python H2 client, such as https://hyper.readthedocs.io/en/latest/, which we already require anyways ?

**Comment by zizhong** [comment]
The twisted code is needed for detecting the goaway frame, which I don't
think cURL or hyper gives us.

Leif Hedstrom <notifications@github.com> 于2019年3月28日周四 上午10:03写道：

**Pull request closed without merging**